cleanrl/tqc_continuous_action.py

```diff
@@ -0,0 +1,363 @@
+    if args.track:
+        import wandb
+
+        wandb.init(
+            project=args.wandb_project_name,
+            entity=args.wandb_entity,
+            sync_tensorboard=True,
+            config=vars(args),
+            name=run_name,
+            monitor_gym=True,
+            save_code=True,
+        )
+    writer = SummaryWriter(f"runs/{run_name}")
+    writer.add_text(
+        "hyperparameters",
+        "|param|value|\n|-|-|\n%s" % ("\n".join([f"|{key}|{value}|" for key, value in vars(args).items()])),
+    )
+
+    # TRY NOT TO MODIFY: seeding
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+    torch.manual_seed(args.seed)
+    torch.backends.cudnn.deterministic = args.torch_deterministic
+
+    device = torch.device("cuda" if torch.cuda.is_available() and args.cuda else "cpu")
+
+    # env setup
+    envs = gym.vector.SyncVectorEnv([make_env(args.env_id, args.seed, 0, args.capture_video, run_name)])
+    assert isinstance(envs.single_action_space, gym.spaces.Box), "only continuous action space is supported"
+
+    max_action = float(envs.single_action_space.high[0])
+
+    if args.use_paper_n_quantiles_to_drop and args.env_id in PAPER_N_QUANTILES_TO_DROP:
+        args.n_quantiles_to_drop = PAPER_N_QUANTILES_TO_DROP[args.env_id]
+        print(f"Using paper n_quantiles_to_drop: {args.n_quantiles_to_drop} for env: {args.env_id}")
+    actor = Actor(envs).to(device)
+    critics = QuantileCritics(envs, args.n_quantiles, args.n_critics).to(device)
+    args.n_top_quantiles_to_drop = args.n_quantiles_to_drop * critics.n_critics
+    actor_optimizer = torch.optim.Adam(actor.parameters(), lr=args.actor_adam_lr)
+    critic_optimizer = torch.optim.Adam(critics.parameters(), lr=args.critic_adam_lr)
+
+    target_critics = QuantileCritics(envs, args.n_quantiles, args.n_critics).to(device)
+    target_critics.load_state_dict(critics.state_dict())
+    target_critics.requires_grad_(False)
+
+    # Automatic entropy tuning
+    target_entropy = -torch.prod(torch.Tensor(envs.single_action_space.shape).to(device)).item()
+    log_alpha = torch.zeros(1, requires_grad=True, device=device)
+    alpha_optimizer = optim.Adam([log_alpha], lr=args.alpha_adam_lr)
+
+    envs.single_observation_space.dtype = np.float32
+    rb = ReplayBuffer(
+        args.buffer_size,
+        envs.single_observation_space,
+        envs.single_action_space,
+        device,
+        handle_timeout_termination=args.handle_timeout_termination,
+    )
+    start_time = time.time()
+
+    # TRY NOT TO MODIFY: start the game
+    obs = envs.reset()
+    for global_step in range(args.total_timesteps):
+        # ALGO LOGIC: put action logic here
+        if global_step < args.learning_starts:
+            actions = np.array([envs.single_action_space.sample() for _ in range(envs.num_envs)])
+        else:
+            actions, _, _ = actor.get_action(torch.Tensor(obs).to(device))
```

Comment:
I suggest enclosing this block with `with torch.no_grad():`

Reply:
Done.